File: fl_mlkit_text_recognize/ios/Classes/FlMlKitTextRecognizeMethodCall.swift
import AVFoundation
import fl_camera
import Flutter
import Foundation
import MLKitTextRecognition
import MLKitTextRecognitionChinese
import MLKitTextRecognitionDevanagari
import MLKitTextRecognitionJapanese
import MLKitTextRecognitionKorean
import MLKitVision

class FlMlKitTextRecognizeMethodCall: FlCameraMethodCall {
    private var options: CommonTextRecognizerOptions = TextRecognizerOptions()
    private var canScan: Bool = false
    private var frequency: Double = 0
    private var lastCurrentTime: TimeInterval = 0
    private var recognizer: TextRecognizer?

    override init(_ _registrar: FlutterPluginRegistrar) {
        super.init(_registrar)
    }

    override func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startPreview":
            frequency = (call.arguments as! [String: Any?])["frequency"] as! Double
            startPreview({ [self] sampleBuffer in
                let currentTime = Date().timeIntervalSince1970 * 1000
                if currentTime - lastCurrentTime >= frequency, canScan {
                    let buffer = CMSampleBufferGetImageBuffer(sampleBuffer)
                    analysis(buffer!.image, nil)
                    lastCurrentTime = currentTime
                }
            }, call: call, result)
        case "setRecognizedLanguage":
            setRecognizedLanguage(call)
            recognizer = nil
            result(true)
        case "scanImageByte":
            let arguments = call.arguments as! [AnyHashable: Any?]
            let useEvent = arguments["useEvent"] as! Bool
            let uint8list = arguments["byte"] as! FlutterStandardTypedData?
            if uint8list != nil {
                let image = UIImage(data: uint8list!.data)
                if image != nil {
                    analysis(image!, useEvent ? nil : result)
                    return
                }
            }
            result([])
        case "scan":
            canScan = call.arguments as! Bool
            result(true)
        case "dispose":
            dispose()
            result(true)
        default:
            super.handle(call: call, result: result)
        }
    }

    override func dispose() {
        super.dispose()
        recognizer = nil
    }

    private func setRecognizedLanguage(_ call: FlutterMethodCall) {
        let type = call.arguments as! String
        switch type {
        case "latin":
            options = TextRecognizerOptions()
        case "chinese":
            options = ChineseTextRecognizerOptions()
        case "japanese":
            options = JapaneseTextRecognizerOptions()
        case "korean":
            options = KoreanTextRecognizerOptions()
        case "devanagari":
            options = DevanagariTextRecognizerOptions()
        default:
            options = TextRecognizerOptions()
        }
    }

    private func getTextRecognition() -> TextRecognizer {
        if recognizer == nil {
            recognizer = TextRecognizer.textRecognizer(options: options)
        }
        return recognizer!
    }

    private func analysis(_ image: UIImage, _ result: FlutterResult?) {
        let visionImage = VisionImage(image: image)
        if flCamera == nil {
            visionImage.orientation = .up
        } else {
            visionImage.orientation = flCamera!.imageOrientation()
        }
        getTextRecognition().process(visionImage) { [self] visionText, error in
            if error == nil, visionText != nil {
                var map = visionText!.data
                map.updateValue(image.size.height, forKey: "height")
                map.updateValue(image.size.width, forKey: "width")
                if result == nil {
                    flCameraEvent?.sendEvent(map)
                } else {
                    result!(map)
                }
            }
        }
    }
}

extension CVBuffer {
    var image: UIImage {
        let ciImage = CIImage(cvPixelBuffer: self)
        let cgImage = CIContext().createCGImage(ciImage, from: ciImage.extent)
        return UIImage(cgImage: cgImage!)
    }

    var image1: UIImage {
        // Lock the base address of the pixel buffer
        CVPixelBufferLockBaseAddress(self, CVPixelBufferLockFlags.readOnly)
        // Get the number of bytes per row for the pixel buffer
        let baseAddress = CVPixelBufferGetBaseAddress(self)
        // Get the number of bytes per row for the pixel buffer
        let bytesPerRow = CVPixelBufferGetBytesPerRow(self)
        // Get the pixel buffer width and height
        let width = CVPixelBufferGetWidth(self)
        let height = CVPixelBufferGetHeight(self)
        // Create a device-dependent RGB color space
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        // Create a bitmap graphics context with the sample buffer data
        var bitmapInfo = CGBitmapInfo.byteOrder32Little.rawValue
        bitmapInfo |= CGImageAlphaInfo.premultipliedFirst.rawValue & CGBitmapInfo.alphaInfoMask.rawValue
        // let bitmapInfo: UInt32 = CGBitmapInfo.alphaInfoMask.rawValue
        let context = CGContext(data: baseAddress, width: width, height: height, bitsPerComponent: 8, bytesPerRow: bytesPerRow, space: colorSpace, bitmapInfo: bitmapInfo)
        // Create a Quartz image from the pixel data in the bitmap graphics context
        let quartzImage = context?.makeImage()
        // Unlock the pixel buffer
        CVPixelBufferUnlockBaseAddress(self, CVPixelBufferLockFlags.readOnly)
        // Create an image object from the Quartz image
        return UIImage(cgImage: quartzImage!)
    }
}

extension CGRect {
    var data: [String: Any?] {
        [
            "x": origin.x,
            "y": origin.y,
            "width": width,
            "height": height,
        ]
    }
}

extension Text {
    var data: [String: Any?] {
        ["text": text,
         "textBlocks": blocks.map {
             $0.data
         }]
    }
}

extension TextBlock {
    var data: [String: Any?] {
        ["text": text,
         "recognizedLanguages": recognizedLanguages.map {
             $0.languageCode
         },
         "boundingBox": frame.data,
         "lines": lines.map {
             $0.data
         },
         "corners": cornerPoints.map {
             $0.cgPointValue.data
         }]
    }
}

extension TextLine {
    var data: [String: Any?] {
        ["text": text,
         "recognizedLanguages": recognizedLanguages.map {
             $0.languageCode
         },
         "elements": elements.map {
             $0.data
         },
         "boundingBox": frame.data,
         "corners": cornerPoints.map {
             $0.cgPointValue.data
         }]
    }
}

extension TextElement {
    var data: [String: Any?] {
        ["text": text,
         "boundingBox": frame.data,
         "corners": cornerPoints.map {
             $0.cgPointValue.data
         }]
    }
}

extension CGPoint {
    var data: [String: Any?] {
        ["x": NSNumber(value: x.native), "y": NSNumber(value: y.native)]
    }
}

File: fl_mlkit_text_recognize/android/src/main/kotlin/fl/mlkit/text/recognize/FlMlKitTextRecognizeMethodCall.kt
package fl.mlkit.text.recognize

import android.annotation.SuppressLint
import android.app.Activity
import android.graphics.BitmapFactory
import android.graphics.Point
import android.graphics.Rect
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.TextRecognizer
import com.google.mlkit.vision.text.TextRecognizerOptionsInterface
import com.google.mlkit.vision.text.chinese.ChineseTextRecognizerOptions
import com.google.mlkit.vision.text.devanagari.DevanagariTextRecognizerOptions
import com.google.mlkit.vision.text.japanese.JapaneseTextRecognizerOptions
import com.google.mlkit.vision.text.korean.KoreanTextRecognizerOptions
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import fl.camera.FlCameraMethodCall
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class FlMlKitTextRecognizeMethodCall(
    activity: Activity,
    plugin: FlutterPlugin.FlutterPluginBinding
) :
    FlCameraMethodCall(activity, plugin) {

    private var options: TextRecognizerOptionsInterface = TextRecognizerOptions.DEFAULT_OPTIONS
    private var recognizer: TextRecognizer? = null
    private var canScan = false
    private var frequency = 0L
    private var lastCurrentTime = 0L

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startPreview" -> {
                val value = call.argument<Double>("frequency")!!
                frequency = value.toLong()
                startPreview(imageAnalyzer, call, result)
            }
            "scanImageByte" -> scanImageByte(call, result)
            "setRecognizedLanguage" -> {
                setRecognizedLanguage(call)
                recognizer?.close()
                recognizer = null
                result.success(true)
            }
            "scan" -> {
                canScan = call.arguments as Boolean
                result.success(true)
            }
            "dispose" -> {
                dispose()
                result.success(true)
            }
            else -> {
                super.onMethodCall(call, result)
            }
        }
    }

    override fun dispose() {
        super.dispose();
        recognizer?.close()
        recognizer = null
    }


    private fun setRecognizedLanguage(call: MethodCall) {
        when (call.arguments as String) {
            "latin" -> options = TextRecognizerOptions.DEFAULT_OPTIONS
            "chinese" -> options =
                ChineseTextRecognizerOptions.Builder().build()
            "japanese" -> options =
                JapaneseTextRecognizerOptions.Builder().build()
            "korean" -> options =
                KoreanTextRecognizerOptions.Builder().build()
            "devanagari" -> options =
                DevanagariTextRecognizerOptions.Builder().build()
        }
    }

    private fun scanImageByte(call: MethodCall, result: MethodChannel.Result) {
        val useEvent = call.argument<Boolean>("useEvent")!!
        val byteArray = call.argument<ByteArray>("byte")!!
        var rotationDegrees = call.argument<Int>("rotationDegrees")
        val bitmap = BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
        if (bitmap == null) {
            result.success(null)
            return
        }
        if (rotationDegrees == null) rotationDegrees = 0
        val inputImage = InputImage.fromBitmap(bitmap, rotationDegrees)
        analysis(inputImage, if (useEvent) null else result, null)
    }

    @SuppressLint("UnsafeOptInUsageError")
    private val imageAnalyzer = ImageAnalysis.Analyzer { imageProxy ->
        val mediaImage = imageProxy.image
        val currentTime = System.currentTimeMillis()
        if (currentTime - lastCurrentTime >= frequency && mediaImage != null && canScan) {
            val inputImage =
                InputImage.fromMediaImage(
                    mediaImage,
                    imageProxy.imageInfo.rotationDegrees
                )
            analysis(inputImage, null, imageProxy)
            lastCurrentTime = currentTime
        } else {
            imageProxy.close()
        }
    }


    private fun analysis(
        inputImage: InputImage,
        result: MethodChannel.Result?,
        imageProxy: ImageProxy?
    ) {
        getTextRecognition().process(inputImage)
            .addOnSuccessListener { visionText ->
                var width = inputImage.width
                var height = inputImage.height
                if (width > height) {
                    width -= height
                    height += width
                    width = height - width
                }
                val map: MutableMap<String, Any?> = HashMap()
                map.putAll(visionText.data)
                map["width"] = width.toDouble()
                map["height"] = height.toDouble()
                if (result == null) {
                    flCameraEvent?.sendEvent(map)
                } else {
                    result.success(map)
                }
            }
            .addOnFailureListener { result?.success(null) }
            .addOnCompleteListener { imageProxy?.close() }

    }

    private fun getTextRecognition(): TextRecognizer {
        if (recognizer == null) {
            recognizer = TextRecognition.getClient(options)
        }
        return recognizer!!
    }


    private val Text.data: Map<String, Any?>
        get() = mapOf(
            "text" to text,
            "textBlocks" to textBlocks.map { textBlock -> textBlock.data }
        )

    private val Text.TextBlock.data: Map<String, Any?>
        get() = mapOf(
            "text" to text,
            "recognizedLanguage" to recognizedLanguage,
            "boundingBox" to boundingBox?.data,
            "corners" to cornerPoints?.map { corner -> corner.data },
            "lines" to lines.map { line -> line.data },
        )
    private val Text.Element.data: Map<String, Any?>
        get() = mapOf(
            "text" to text,
            "recognizedLanguage" to recognizedLanguage,
            "boundingBox" to boundingBox?.data,
            "corners" to cornerPoints?.map { corner -> corner.data },
        )
    private val Text.Line.data: Map<String, Any?>
        get() = mapOf(
            "text" to text,
            "recognizedLanguage" to recognizedLanguage,
            "boundingBox" to boundingBox?.data,
            "corners" to cornerPoints?.map { corner -> corner.data },
            "elements" to elements.map { element -> element.data },
        )

    private val Rect.data: Map<String, Int>
        get() = mapOf(
            "top" to top,
            "bottom" to bottom,
            "left" to left,
            "right" to right
        )

    private val Point.data: Map<String, Double>
        get() = mapOf("x" to x.toDouble(), "y" to y.toDouble())
}
